### PR TITLE
Address leak due to RandomNumberControl::Children

### PR DESCRIPTION
**Pull request by williamfgc**
Replace raw pointer with unique_ptr

Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Replacing vector of raw pointers with `std::unique_ptr` for objects of type `RandomGenerator_t` for static RandomNumberControl::Children. Also, use appropriate ownership transfer `reset` and `release` where appropriate. This solves an identified leak with asan/lsan as part of #3083.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix (leak)
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Ubuntu 20.04. Must pass CI.

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by williamfgc** [comment]
please disregard the `codecov/patch` check, the important one is `codecov/project`

File: src/OhmmsApp/RandomNumberControl.cpp [hunk 1]
```diff
@@ -258,8 +259,8 @@ void RandomNumberControl::read_old(const std::string& fname, Communicate* comm)
     hout.push("random");
     std::string engname;
     hout.read(slab, Random.EngineName);
-    shape[0] = static_cast<int>(slab.size(0));
-    shape[1] = static_cast<int>(slab.size(1));
+    shape[0]           = static_cast<int>(slab.size(0));
```
Comment (ye-luo) [inline src/OhmmsApp/RandomNumberControl.cpp:262]: Looks like a bug in clang-format. No worth fixing the formatting on our side but just wait.

**Review by ye-luo — APPROVED** [review]
Good cleaning. Thanks.

File: src/OhmmsApp/RandomNumberControl.cpp (hunk 1 above)
Comment (williamfgc) [reply src/OhmmsApp/RandomNumberControl.cpp:262]: @ye-luo Yes, thanks for pointing it out. Let's discuss on slack as we can add a static check for clang-format and settle on a specific version (I'm using 10, but whichever makes more sense for everybody). My IDE has automatic formatting hence there is a few of these. In the meantime, let me undo these clang-format related changes.

Comment (ye-luo) [reply src/OhmmsApp/RandomNumberControl.cpp:262]: I approved the PR, namely I would like to keep these formatting. I only encourage people not formatting the source code when a PR touch a lot of files and complicates the review.

Comment (williamfgc) [reply src/OhmmsApp/RandomNumberControl.cpp:262]: OK, sounds good!

**Comment by ye-luo** [comment]
Okay to test

**Pull request merged**